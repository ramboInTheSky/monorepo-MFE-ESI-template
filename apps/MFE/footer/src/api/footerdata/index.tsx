import logger from "@monorepo/core-logger"
import {IncomingHttpHeaders} from "http"
import axios from "../../utils/axios"

import {FooterModel} from "../../models/footerModel"
import FooterApiConfig from "../../config/api/footerdata"

// this will always be called server-side on first render and subsequent server-side renders
export const getFooterData = async (headers: IncomingHttpHeaders): Promise<FooterModel> => {
    try {
        const getFooterDataApi = FooterApiConfig("getFooterData")
        const response = await axios({
            method: getFooterDataApi.method,
            url: `${getFooterDataApi.localEndpoint()}`,
            headers,
        })

        return response.data
    } catch (ex) {
        logger.error(ex)
        throw ex
    }
}
