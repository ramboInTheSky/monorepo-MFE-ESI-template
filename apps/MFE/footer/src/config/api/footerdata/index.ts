import {Api, HTTPMethod, buildExport} from "@monorepo/apiconfig"
import env from "../../env"

const {REACT_APP_API_BASEURL, REACT_APP_ENTRYPOINT} = env

export class FooterApi implements Api {
    baseURL = `${REACT_APP_API_BASEURL}${REACT_APP_ENTRYPOINT}`
    version = "v1"
    endpoints = {
        getFooterData: {
            routeDefinition: "/footerdata",
            getLocalURL: () => `/footerdata`,
            getRemoteURL: () => `/footers/Default`,
            method: HTTPMethod.get,
        },
    }
}

const footerApi = new FooterApi()
export default (endpoint: keyof typeof footerApi.endpoints) => buildExport(footerApi, endpoint)
