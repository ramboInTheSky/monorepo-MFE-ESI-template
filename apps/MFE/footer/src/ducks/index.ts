import {combineReducers, createStore, applyMiddleware} from "redux"
import {composeWithDevTools} from "redux-devtools-extension"
import thunk from "redux-thunk"
import requestReducer, {RequestDuckState} from "./request"
import env from "../config/env"
import footerReducer, {FooterDuckState} from "./footerdata"
import userReducer, {UserDuckState} from "./user"
import textReducer, {TextDuckState} from "./text"
import languageReducer, {LanguageDuckState} from "./languages"
import textAlignmentReducer from "./text-alignment"
import settingsReducer, {SettingsDuckState} from "./settings"

const {DEVELOPMENT} = env

export interface State {
    user: UserDuckState
    request: RequestDuckState
    languages: LanguageDuckState
    data: FooterDuckState
    textAlignment: string
    text: TextDuckState
    settings: SettingsDuckState
}

export const rootReducer = combineReducers({
    user: userReducer,
    request: requestReducer,
    languages: languageReducer,
    data: footerReducer,
    textAlignment: textAlignmentReducer,
    text: textReducer,
    settings: settingsReducer,
})

export const makeStore = (initialState?: State) => {
    return createStore(
        rootReducer,
        initialState,
        DEVELOPMENT ? composeWithDevTools(applyMiddleware(thunk)) : applyMiddleware(thunk),
    )
}
