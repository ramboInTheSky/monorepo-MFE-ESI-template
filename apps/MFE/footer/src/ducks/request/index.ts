import {getSettingsHeaders} from "@monorepo/utils"
import {IncomingHttpHeaders} from "http"
import {Store} from "redux"
import {COUNTRY_GB, TERRITORY_HEADER} from "../../config/constants"

export const SET_REQUEST = "SET_REQUEST"

interface SetRequest {
    type: typeof SET_REQUEST
    request: any
}

export interface RequestDuckState {
    headers: IncomingHttpHeaders | null
    url: string | null
    isInternationalCountry: boolean
}

const initialState: RequestDuckState = {
    headers: null,
    url: null,
    isInternationalCountry: false,
}

export const setRequestAction = (store: Store, request: any) => {
    store.dispatch({
        type: SET_REQUEST,
        request,
    })
}

const reducer = (state: RequestDuckState = initialState, action: SetRequest): RequestDuckState => {
    const {request} = action

    switch (action.type) {
        case SET_REQUEST:
            return {
                ...state,
                headers: getSettingsHeaders(request.headers),
                url: request.url,
                isInternationalCountry:
                    request.headers![TERRITORY_HEADER] &&
                    request.headers![TERRITORY_HEADER].toLowerCase() !== COUNTRY_GB,
            }
        default:
            return state
    }
}

export default reducer
