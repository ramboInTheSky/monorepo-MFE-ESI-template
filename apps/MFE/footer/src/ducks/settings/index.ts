import {VARIANT_DEFAULT} from "../../config/constants"

export const SET_SETTINGS = "SET_SETTINGS"

export interface SettingsDuckState {
    variant: string
}

interface SetSettings {
    type: typeof SET_SETTINGS
    value: SettingsDuckState
}

type SettingsActions = SetSettings

const initalState: SettingsDuckState = {
    variant: VARIANT_DEFAULT,
}

export const setSettings = (value: SettingsDuckState): SetSettings => ({
    type: SET_SETTINGS,
    value,
})

export const updateSettings = (store, settings) => {
    store.dispatch(setSettings(settings))
}

export const settingsReducer = (state: SettingsDuckState = initalState, action: SettingsActions): SettingsDuckState => {
    switch (action.type) {
        case SET_SETTINGS:
            return {
                ...state,
                ...action.value,
            }

        default:
            return state
    }
}

export default settingsReducer
