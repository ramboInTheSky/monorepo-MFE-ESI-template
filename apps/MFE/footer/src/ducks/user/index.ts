import { Store } from 'redux'

export const SET_USER = 'SET_USER'

interface SetUser {
  type: typeof SET_USER;
  user: UserDuckState;
}

export interface AccountStatusChanged {
  (accountFirstName: string): SetUser;
}

export interface UserDuckState {
  accountFirstName: string;
  loggedIn: boolean;
}

const initialState: UserDuckState = { accountFirstName: '', loggedIn: false }

export const setUser = (user: UserDuckState): SetUser => ({
  type: SET_USER,
  user,
})

export const updateUserStatus = (store: Store, accountFirstName: string) => {
  store.dispatch(setUser({ accountFirstName, loggedIn: !!accountFirstName }))
}

const reducer = (
  state: UserDuckState = initialState,
  action: SetUser,
): UserDuckState => {
  switch (action.type) {
    case SET_USER:
      return { ...action.user }
    default:
      return state
  }
}

export default reducer
