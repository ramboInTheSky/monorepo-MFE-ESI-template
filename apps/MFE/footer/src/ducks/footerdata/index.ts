import logger from "@monorepo/core-logger"
import {Store} from "redux"
import {FooterModel} from "../../models/footerModel"
import {getFooterData} from "../../api/footerdata"

export const SET_FOOTER_DATA = "SET_FOOTER_DATA"

interface SetFooterData {
    type: typeof SET_FOOTER_DATA
    data: FooterModel
}

export type FooterDuckState = FooterModel | null

const initialState: FooterDuckState = null

const setFooterDataAction = (data: FooterModel) => {
    return {
        type: SET_FOOTER_DATA,
        data,
    }
}

export const getFooterDataThunk = async ({dispatch, getState}: Store) => {
    try {
        const {request} = getState()
        const response = await getFooterData(request.headers!)
        dispatch(setFooterDataAction(response))
    } catch (error) {
        logger.error(error)
        dispatch(setFooterDataAction(new FooterModel()))
    }
}

const reducer = (state: FooterDuckState = initialState, action: SetFooterData): FooterDuckState => {
    const {data} = action

    switch (action.type) {
        case SET_FOOTER_DATA:
            return {
                ...state,
                ...data,
            }
        default:
            return state
    }
}

export default reducer
