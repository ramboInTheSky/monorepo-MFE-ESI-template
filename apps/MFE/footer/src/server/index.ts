/* eslint-disable @typescript-eslint/no-var-requires */
/* istanbul ignore file */
import path from "path"
import express, {Express} from "express"
import {SettingsSDK} from "@monorepo/settings-sdk"
import NodeCache from "node-cache"
import {setup, defaultClient} from "applicationinsights"
import {helmetGuard, correlationMiddleware, httpLogger, siteurlMiddleware, textMiddleware} from "@monorepo/middlewares"
import htmlMiddleware from "./middleware/render/html"
import renderMiddleware from "./middleware/render/render"
import staticContentMiddleware from "./middleware/staticContent"
import themeMiddleware from "./middleware/theme"
import containerMiddleware from "./middleware/container"
import CacheControlMiddleware, {StaticAssetsCacheControlMiddleware} from "./middleware/cache-control"
import AppInsightsMiddleware from "./middleware/appInsights"
import apis from "./apis"
import env from "../config/env"
import cssMiddleware from "./middleware/render/css"
import {PRELOAD, BUNDLE_VENDOR, PRELOAD_JSS_VENDOR, BUNDLE, PRELOAD_VENDOR} from "./middleware/render/assets"
import {SettingsSdkKeys} from "../models/settings"
import BFFLogger from "./core/BFFLogger"

const appRoutes = `${env.REACT_APP_ENTRYPOINT}`
const cache = new NodeCache()
const publicPath = path.join(__dirname, "/public")
const app: any = express()

app.use(correlationMiddleware)

app.use(siteurlMiddleware(env.ASSETS_PATH, env.DEV_URL_OVERRIDE))

if (env.REACT_APP_APPINSIGHTS_KEY) {
    setup(env.REACT_APP_APPINSIGHTS_KEY).start()
    app.use(AppInsightsMiddleware(defaultClient))
}

app.use(httpLogger(env.DEVELOPMENT))

app.use(helmetGuard)

app.get("/", (_req, res) => {
    res.status(200).send("UP")
})

app.use(
    `${env.REACT_APP_BLOB_STORAGE_PATH}/footer`,
    StaticAssetsCacheControlMiddleware,
    express.static(`${publicPath}/static`),
)
app.use(staticContentMiddleware)
app.use(env.ASSETS_PATH, StaticAssetsCacheControlMiddleware, express.static(publicPath))
app.use(
    `${env.REACT_APP_BLOB_STORAGE_PATH}`,
    StaticAssetsCacheControlMiddleware,
    express.static(`${publicPath}/static-content`),
)

app.use("/footercontainer", containerMiddleware)

app.use(SettingsSDK("MONOREPO.FOOTER.FRONTEND", "./featuresettings.json"))

app.use(textMiddleware(publicPath, SettingsSdkKeys, cache, BFFLogger))
app.use(themeMiddleware(cache))


app.get(`${env.ASSETS_PATH}/dependencies/:type`, CacheControlMiddleware, (_req, res) => {
    const versionTheme = res.locals.configuration[SettingsSdkKeys.themeVersion]?.Value
    const SITEURL = _req.siteUrl.url
    try {
        const {theme} = _req as any
        if (_req.params.type === "preloadjssvendor") {
            res.status(200).send(PRELOAD_JSS_VENDOR(theme, versionTheme))
        } else if (_req.params.type === "preloadvendor") {
            res.status(200).send(PRELOAD_VENDOR(theme, versionTheme))
        } else if (_req.params.type === "preload") {
            res.status(200).send(PRELOAD(theme, versionTheme))
        } else if (_req.params.type === "bundlevendor") {
            res.status(200).send(BUNDLE_VENDOR(SITEURL))
        } else {
            res.status(200).send(BUNDLE(SITEURL))
        }
    } catch (err) {
        BFFLogger.error(err)
    }
})

app.use(apis)

app.get(`${env.ASSETS_PATH}/baselinecss/:type`, CacheControlMiddleware, cssMiddleware)

app.get(`${appRoutes}/health`, (_req, res) => {
    // I have no dependencies to check health upon
    res.status(200).send("Healthy")
})

app.use(`${appRoutes}/flush`, (_req, res) => {
    try {
        cache.flushAll()
        res.status(200).end()
    } catch {
        res.status(500).end()
    }
})

app.use(appRoutes, htmlMiddleware())
app.use(appRoutes, renderMiddleware, CacheControlMiddleware, (_req, _res) => {
    _res.status(200).send((_res as any).html)
})

export default app
