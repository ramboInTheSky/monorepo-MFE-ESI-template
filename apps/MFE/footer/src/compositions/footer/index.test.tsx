import React from "react"
import {render} from "@testing-library/react"

import BFFLogger from "../../server/core/BFFLogger"
import mockStore, {mockState} from "../../../__mocks__/mockStore"

import {mockDateConstructor} from "../../../__mocks__/setDate"
import {getFooterDataThunk} from "../../ducks/footerdata"

import {Footer} from "."
import {getServerSideProps} from "./index.server"
import apiFooterData from "../../../__mocks__/apiFooterData"

jest.mock("../../components/RegionWrapper", () => ({
    __esModule: true,
    // eslint-disable-next-line react/display-name
    default: ({regionType}) => {
        return <div>TEST REGION WRAPPER {regionType}</div>
    },
}))
jest.mock("../../components/Copyright/DefaultCopyright", () => ({
    __esModule: true,
    // eslint-disable-next-line react/display-name
    default: () => {
        return <div>DEFAULT COPYRIGHT</div>
    },
}))

jest.mock("@app/utils/axios")

jest.mock("@app/utils/setApiUrlSettings", () => {
    const mock = jest.fn()
    mock.mockReturnValue({
        realm: "testRealm",
        territory: "testTerritory",
        language: "testLanguage",
    })
    return {
        setApiUrlSettings: mock,
    }
})

jest.mock("../../ducks/footerdata", () => ({
    getFooterDataThunk: jest.fn(store => {
        if (store.error) throw new Error("ERROR")

        return {}
    }),
}))
jest.mock("../../server/core/BFFLogger", () => ({
    error: jest.fn(),
    warn: jest.fn(),
}))

describe("Pages: Footer - ", () => {
    describe("When I try provide the data to the footer and an error has occured: ", () => {
        beforeEach(() => {
            mockDateConstructor(new Date("2019-12-08T07:00:00.000Z"))
        })
        it("should show the default footer", () => {
            const {asFragment} = render(<Footer {...{data: null, textAlignment: "ltr"}} />)
            expect(asFragment()).toMatchSnapshot()
        })
    })

    describe("When the footer renders with regions ", () => {
        it("should show the default footer", () => {
            const {asFragment} = render(
                <Footer {...{data: apiFooterData, regions: mockState.data?.regions, textAlignment: "ltr"}} />,
            )
            expect(asFragment()).toMatchSnapshot()
        })
    })

    describe("When the footer getServerSideProps function is called ", () => {
        let props: any
        beforeAll(async () => {
            props = await getServerSideProps({params: {}}, {}, mockStore)
        })
        it("should call GetFooterData", () => {
            expect(getFooterDataThunk).toHaveBeenCalledWith(mockStore)
        })

        it("should return data from the API", () => {
            expect(props).toBeTruthy()
        })
        it("should return correct values", () => {
            expect(props).toEqual({})
        })
    })

    describe("When the footer getServerSideProps throws an error ", () => {
        let props: any
        beforeAll(async () => {
            props = await getServerSideProps({}, {}, {error: true} as any)
        })

        it("should call the error logger", () => {
            // eslint-disable-next-line @typescript-eslint/unbound-method
            expect(BFFLogger.error).toHaveBeenCalledWith(expect.any(Error))
        })

        it("should return data from the API", () => {
            expect(props).toBeTruthy()
        })
        it("should return correct values", () => {
            expect(props).toEqual({})
        })
    })
})
