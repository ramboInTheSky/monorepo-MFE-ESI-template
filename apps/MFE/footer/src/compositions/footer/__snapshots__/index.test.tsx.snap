// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages: Footer -  When I try provide the data to the footer and an error has occured:  should show the default footer 1`] = `
<DocumentFragment>
  <footer
    dir="ltr"
  >
    <div>
      DEFAULT COPYRIGHT
    </div>
  </footer>
</DocumentFragment>
`;

exports[`Pages: Footer -  When the footer renders with regions  should show the default footer 1`] = `
<DocumentFragment>
  <footer
    dir="ltr"
  >
    <div
      data-testid="footer-container"
    >
      <div>
        TEST REGION WRAPPER SocialMedia
      </div>
      <div>
        TEST REGION WRAPPER QuickLinks
      </div>
      <div>
        TEST REGION WRAPPER MainLinks
      </div>
      <div>
        TEST REGION WRAPPER Copyright
      </div>
    </div>
  </footer>
</DocumentFragment>
`;
